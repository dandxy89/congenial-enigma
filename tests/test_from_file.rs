use std::path::PathBuf;

use lp_parser_rs::{
    model::{lp_error::LPParserError, lp_problem::LPProblem},
    parse::{parse_file, parse_lp_file},
};

#[macro_export]
macro_rules! generate_test {
    ($test_name:ident, $file:expr, $o_sum:expr, $c_sum:expr) => {
        #[test]
        fn $test_name() {
            let result = read_file_from_resources($file).unwrap();

            #[cfg(feature = "serde")]
            insta::assert_yaml_snapshot!(result, {
                ".variables" => insta::sorted_redaction(),
                ".objectives" => insta::sorted_redaction(),
                ".constraints" => insta::sorted_redaction(),
            });

            let summation = &result.objectives.iter().map(|o| o.coefficients.iter().map(|c| c.coefficient).sum::<f64>()).sum();
            float_eq::assert_float_eq!($o_sum, *summation, abs <= 1e-2);

            let summation = &result.constraints.iter().map(|(_, c)| c.coefficients().iter().map(|c| c.coefficient).sum::<f64>()).sum();
            float_eq::assert_float_eq!($c_sum, *summation, abs <= 1e-2);
        }
    };
}

generate_test!(afiro, "afiro.lp", -4.100, 25.369);
generate_test!(afiro_ext, "afiro_ext.lp", -5.239, 25.369);
generate_test!(boeing1, "boeing1.lp", 1187.985, 194612.346);
generate_test!(boeing2, "boeing2.lp", 78.488, 20863.836);
generate_test!(fit1d, "fit1d.lp", 82457.0, -146871.180);
generate_test!(kb2, "kb2.lp", 11.675, 10143.724);
generate_test!(pulp, "pulp.lp", 1., 73.44);
generate_test!(pulp2, "pulp2.lp", -6.0, 147.);
generate_test!(sc50a, "sc50a.lp", -1., 30.3);
generate_test!(model2, "model2.lp", 0., 6.);
generate_test!(limbo, "limbo.lp", 2., 0.);
generate_test!(obj3_2cons, "3obj_2cons.lp", 1., -186.);
generate_test!(obj_2cons_only_binary_vars, "2obj_2cons_only_binary_vars.lp", -7.5, -186.);
generate_test!(obj_2cons_all_variable_types, "2obj_2cons_all_variable_types.lp", -7.5, -186.);
generate_test!(obj_1cons_all_variables_with_bounds, "1obj_1cons_all_variables_with_bounds.lp", -1., 16.5);
generate_test!(semi_continuous, "semi_continuous.lp", 2., 0.);
generate_test!(sos, "sos.lp", 0., 17.5);
generate_test!(test, "test.lp", 2., 2.9899);
generate_test!(test2, "test2.lp", -6., 147.);
generate_test!(empty_bounds, "empty_bounds.lp", 11., 2.);
generate_test!(blank_lines, "blank_lines.lp", 11., 2.);
generate_test!(optional_labels, "optional_labels.lp", 11., 1.);
generate_test!(infile_comments, "infile_comments.lp", 43., 7.);
generate_test!(infile_comments2, "infile_comments2.lp", 43., 0.);
generate_test!(missing_signs, "missing_signs.lp", 43., -3.);
generate_test!(fit2d, "fit2d.lp", 349048.9, -296677.389);
generate_test!(scientific_notation, "scientific_notation.lp", 238.0324, -43.87680);

#[test]
fn invalid() {
    let result = read_file_from_resources("invalid.lp");
    assert!(result.is_err());
}

fn read_file_from_resources(file_name: &str) -> Result<LPProblem, LPParserError> {
    let mut file_path = PathBuf::from(env!("CARGO_MANIFEST_DIR"));
    file_path.push(format!("resources/{file_name}"));
    let contents = parse_file(&file_path)?;
    parse_lp_file(&contents)
}
