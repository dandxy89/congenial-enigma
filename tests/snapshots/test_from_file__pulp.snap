---
source: tests/test_from_file.rs
expression: result
---
problem_name: ~
problem_sense: Minimize
variables:
  A_0: General
  A_3: General
  A_7: General
  B_0: General
  B_3: General
  B_7: General
  C_0: General
  C_7: General
  D_0: General
  D_3: General
  N_0: General
  N_3: General
  N_7: General
  PenaltyTree_48: General
  over_0: General
  over_10: General
  over_12: General
  over_14: General
  over_16: General
  over_18: General
  over_2: General
  over_20: General
  over_22: General
  over_24: General
  over_26: General
  over_28: General
  over_30: General
  over_32: General
  over_34: General
  over_36: General
  over_38: General
  over_4: General
  over_40: General
  over_42: General
  over_44: General
  over_46: General
  over_6: General
  over_8: General
  under_1: General
  under_11: General
  under_13: General
  under_15: General
  under_17: General
  under_19: General
  under_21: General
  under_23: General
  under_25: General
  under_27: General
  under_29: General
  under_3: General
  under_31: General
  under_33: General
  under_35: General
  under_37: General
  under_39: General
  under_41: General
  under_43: General
  under_45: General
  under_47: General
  under_5: General
  under_7: General
  under_9: General
objectives:
  - name: OBJ
    coefficients:
      - var_name: PenaltyTree_48
        coefficient: 1
constraints:
  "_C10:":
    Standard:
      name: "_C10:"
      coefficients:
        - var_name: B_3
          coefficient: -1
        - var_name: B_7
          coefficient: 1
        - var_name: under_9
          coefficient: 1
      sense: ">="
      rhs: 0
  "_C11:":
    Standard:
      name: "_C11:"
      coefficients:
        - var_name: D_0
          coefficient: -1
        - var_name: D_3
          coefficient: 1
        - var_name: over_10
          coefficient: -1
      sense: "<="
      rhs: 0
  "_C12:":
    Standard:
      name: "_C12:"
      coefficients:
        - var_name: D_0
          coefficient: -1
        - var_name: D_3
          coefficient: 1
        - var_name: under_11
          coefficient: 1
      sense: ">="
      rhs: 0
  "_C13:":
    Standard:
      name: "_C13:"
      coefficients:
        - var_name: N_0
          coefficient: -1
        - var_name: N_3
          coefficient: 1
        - var_name: over_12
          coefficient: -1
      sense: "<="
      rhs: 0
  "_C14:":
    Standard:
      name: "_C14:"
      coefficients:
        - var_name: N_0
          coefficient: -1
        - var_name: N_3
          coefficient: 1
        - var_name: under_13
          coefficient: 1
      sense: ">="
      rhs: 0
  "_C15:":
    Standard:
      name: "_C15:"
      coefficients:
        - var_name: N_3
          coefficient: -1
        - var_name: N_7
          coefficient: 1
        - var_name: over_14
          coefficient: -1
      sense: "<="
      rhs: 0
  "_C16:":
    Standard:
      name: "_C16:"
      coefficients:
        - var_name: N_3
          coefficient: -1
        - var_name: N_7
          coefficient: 1
        - var_name: under_15
          coefficient: 1
      sense: ">="
      rhs: 0
  "_C17:":
    Standard:
      name: "_C17:"
      coefficients:
        - var_name: A_7
          coefficient: 1
        - var_name: B_7
          coefficient: -1
        - var_name: over_16
          coefficient: -1
      sense: "<="
      rhs: 0
  "_C18:":
    Standard:
      name: "_C18:"
      coefficients:
        - var_name: A_7
          coefficient: 1
        - var_name: B_7
          coefficient: -1
        - var_name: under_17
          coefficient: 1
      sense: ">="
      rhs: 0
  "_C19:":
    Standard:
      name: "_C19:"
      coefficients:
        - var_name: A_7
          coefficient: -1
        - var_name: N_7
          coefficient: 1
        - var_name: over_18
          coefficient: -1
      sense: "<="
      rhs: 0
  "_C1:":
    Standard:
      name: "_C1:"
      coefficients:
        - var_name: A_0
          coefficient: -1
        - var_name: A_3
          coefficient: 1
        - var_name: over_0
          coefficient: -1
      sense: "<="
      rhs: 0
  "_C20:":
    Standard:
      name: "_C20:"
      coefficients:
        - var_name: A_7
          coefficient: -1
        - var_name: N_7
          coefficient: 1
        - var_name: under_19
          coefficient: 1
      sense: ">="
      rhs: 0
  "_C21:":
    Standard:
      name: "_C21:"
      coefficients:
        - var_name: C_7
          coefficient: 1
        - var_name: N_7
          coefficient: -1
        - var_name: over_20
          coefficient: -1
      sense: "<="
      rhs: 0
  "_C22:":
    Standard:
      name: "_C22:"
      coefficients:
        - var_name: C_7
          coefficient: 1
        - var_name: N_7
          coefficient: -1
        - var_name: under_21
          coefficient: 1
      sense: ">="
      rhs: 0
  "_C23:":
    Standard:
      name: "_C23:"
      coefficients:
        - var_name: B_7
          coefficient: 1
        - var_name: C_7
          coefficient: -1
        - var_name: over_22
          coefficient: -1
      sense: "<="
      rhs: 0
  "_C24:":
    Standard:
      name: "_C24:"
      coefficients:
        - var_name: B_7
          coefficient: 1
        - var_name: C_7
          coefficient: -1
        - var_name: under_23
          coefficient: 1
      sense: ">="
      rhs: 0
  "_C25:":
    Standard:
      name: "_C25:"
      coefficients:
        - var_name: B_3
          coefficient: 1
        - var_name: N_3
          coefficient: -1
        - var_name: over_24
          coefficient: -1
      sense: "<="
      rhs: 0
  "_C26:":
    Standard:
      name: "_C26:"
      coefficients:
        - var_name: B_3
          coefficient: 1
        - var_name: N_3
          coefficient: -1
        - var_name: under_25
          coefficient: 1
      sense: ">="
      rhs: 0
  "_C27:":
    Standard:
      name: "_C27:"
      coefficients:
        - var_name: A_3
          coefficient: 1
        - var_name: B_3
          coefficient: -1
        - var_name: over_26
          coefficient: -1
      sense: "<="
      rhs: 0
  "_C28:":
    Standard:
      name: "_C28:"
      coefficients:
        - var_name: A_3
          coefficient: 1
        - var_name: B_3
          coefficient: -1
        - var_name: under_27
          coefficient: 1
      sense: ">="
      rhs: 0
  "_C29:":
    Standard:
      name: "_C29:"
      coefficients:
        - var_name: A_3
          coefficient: -1
        - var_name: N_3
          coefficient: 1
        - var_name: over_28
          coefficient: -1
      sense: "<="
      rhs: 0
  "_C2:":
    Standard:
      name: "_C2:"
      coefficients:
        - var_name: A_0
          coefficient: -1
        - var_name: A_3
          coefficient: 1
        - var_name: under_1
          coefficient: 1
      sense: ">="
      rhs: 0
  "_C30:":
    Standard:
      name: "_C30:"
      coefficients:
        - var_name: A_3
          coefficient: -1
        - var_name: N_3
          coefficient: 1
        - var_name: under_29
          coefficient: 1
      sense: ">="
      rhs: 0
  "_C31:":
    Standard:
      name: "_C31:"
      coefficients:
        - var_name: A_0
          coefficient: 1
        - var_name: N_0
          coefficient: -1
        - var_name: over_30
          coefficient: -1
      sense: "<="
      rhs: 0
  "_C32:":
    Standard:
      name: "_C32:"
      coefficients:
        - var_name: A_0
          coefficient: 1
        - var_name: N_0
          coefficient: -1
        - var_name: under_31
          coefficient: 1
      sense: ">="
      rhs: 0
  "_C33:":
    Standard:
      name: "_C33:"
      coefficients:
        - var_name: A_0
          coefficient: -1
        - var_name: D_0
          coefficient: 1
        - var_name: over_32
          coefficient: -1
      sense: "<="
      rhs: 0
  "_C34:":
    Standard:
      name: "_C34:"
      coefficients:
        - var_name: A_0
          coefficient: -1
        - var_name: D_0
          coefficient: 1
        - var_name: under_33
          coefficient: 1
      sense: ">="
      rhs: 0
  "_C35:":
    Standard:
      name: "_C35:"
      coefficients:
        - var_name: B_0
          coefficient: 1
        - var_name: D_0
          coefficient: -1
        - var_name: over_34
          coefficient: -1
      sense: "<="
      rhs: 0
  "_C36:":
    Standard:
      name: "_C36:"
      coefficients:
        - var_name: B_0
          coefficient: 1
        - var_name: D_0
          coefficient: -1
        - var_name: under_35
          coefficient: 1
      sense: ">="
      rhs: 0
  "_C37:":
    Standard:
      name: "_C37:"
      coefficients:
        - var_name: B_0
          coefficient: -1
        - var_name: N_0
          coefficient: 1
        - var_name: over_36
          coefficient: -1
      sense: "<="
      rhs: 0
  "_C38:":
    Standard:
      name: "_C38:"
      coefficients:
        - var_name: B_0
          coefficient: -1
        - var_name: N_0
          coefficient: 1
        - var_name: under_37
          coefficient: 1
      sense: ">="
      rhs: 0
  "_C39:":
    Standard:
      name: "_C39:"
      coefficients:
        - var_name: D_3
          coefficient: 1
        - var_name: over_38
          coefficient: -1
      sense: "<="
      rhs: 1.24660824696
  "_C3:":
    Standard:
      name: "_C3:"
      coefficients:
        - var_name: A_3
          coefficient: -1
        - var_name: A_7
          coefficient: 1
        - var_name: over_2
          coefficient: -1
      sense: "<="
      rhs: 0
  "_C40:":
    Standard:
      name: "_C40:"
      coefficients:
        - var_name: D_3
          coefficient: 1
        - var_name: under_39
          coefficient: 1
      sense: ">="
      rhs: 1.24660824696
  "_C41:":
    Standard:
      name: "_C41:"
      coefficients:
        - var_name: A_7
          coefficient: 1
        - var_name: B_7
          coefficient: 1
        - var_name: C_7
          coefficient: 1
        - var_name: N_7
          coefficient: 1
        - var_name: over_40
          coefficient: -1
      sense: "<="
      rhs: 9.80665154278
  "_C42:":
    Standard:
      name: "_C42:"
      coefficients:
        - var_name: A_7
          coefficient: 1
        - var_name: B_7
          coefficient: 1
        - var_name: C_7
          coefficient: 1
        - var_name: N_7
          coefficient: 1
        - var_name: under_41
          coefficient: 1
      sense: ">="
      rhs: 9.80665154278
  "_C43:":
    Standard:
      name: "_C43:"
      coefficients:
        - var_name: C_0
          coefficient: 1
        - var_name: over_42
          coefficient: -1
      sense: "<="
      rhs: 2.06580795211
  "_C44:":
    Standard:
      name: "_C44:"
      coefficients:
        - var_name: C_0
          coefficient: 1
        - var_name: under_43
          coefficient: 1
      sense: ">="
      rhs: 2.06580795211
  "_C45:":
    Standard:
      name: "_C45:"
      coefficients:
        - var_name: A_3
          coefficient: 1
        - var_name: B_3
          coefficient: 1
        - var_name: N_3
          coefficient: 1
        - var_name: over_44
          coefficient: -1
      sense: "<="
      rhs: 7.1056670077
  "_C46:":
    Standard:
      name: "_C46:"
      coefficients:
        - var_name: A_3
          coefficient: 1
        - var_name: B_3
          coefficient: 1
        - var_name: N_3
          coefficient: 1
        - var_name: under_45
          coefficient: 1
      sense: ">="
      rhs: 7.1056670077
  "_C47:":
    Standard:
      name: "_C47:"
      coefficients:
        - var_name: A_0
          coefficient: 1
        - var_name: B_0
          coefficient: 1
        - var_name: D_0
          coefficient: 1
        - var_name: N_0
          coefficient: 1
        - var_name: over_46
          coefficient: -1
      sense: "<="
      rhs: 7.89518556411
  "_C48:":
    Standard:
      name: "_C48:"
      coefficients:
        - var_name: A_0
          coefficient: 1
        - var_name: B_0
          coefficient: 1
        - var_name: D_0
          coefficient: 1
        - var_name: N_0
          coefficient: 1
        - var_name: under_47
          coefficient: 1
      sense: ">="
      rhs: 7.89518556411
  "_C49:":
    Standard:
      name: "_C49:"
      coefficients:
        - var_name: PenaltyTree_48
          coefficient: -1
        - var_name: over_0
          coefficient: 0.5
        - var_name: over_10
          coefficient: 0.5
        - var_name: over_12
          coefficient: 0.5
        - var_name: over_14
          coefficient: 0.5
        - var_name: over_16
          coefficient: 0.02
        - var_name: over_18
          coefficient: 0.02
        - var_name: over_2
          coefficient: 0.5
        - var_name: over_20
          coefficient: 0.02
        - var_name: over_22
          coefficient: 0.02
        - var_name: over_24
          coefficient: 0.02
        - var_name: over_26
          coefficient: 0.02
        - var_name: over_28
          coefficient: 0.02
        - var_name: over_30
          coefficient: 0.02
        - var_name: over_32
          coefficient: 0.02
        - var_name: over_34
          coefficient: 0.02
        - var_name: over_36
          coefficient: 0.02
        - var_name: over_38
          coefficient: 4
        - var_name: over_4
          coefficient: 0.5
        - var_name: over_40
          coefficient: 4
        - var_name: over_42
          coefficient: 4
        - var_name: over_44
          coefficient: 4
        - var_name: over_46
          coefficient: 4
        - var_name: over_6
          coefficient: 0.5
        - var_name: over_8
          coefficient: 0.5
        - var_name: under_1
          coefficient: 0.5
        - var_name: under_11
          coefficient: 0.5
        - var_name: under_13
          coefficient: 0.5
        - var_name: under_15
          coefficient: 0.5
        - var_name: under_17
          coefficient: 0.02
        - var_name: under_19
          coefficient: 0.02
        - var_name: under_21
          coefficient: 0.02
        - var_name: under_23
          coefficient: 0.02
        - var_name: under_25
          coefficient: 0.02
        - var_name: under_27
          coefficient: 0.02
        - var_name: under_29
          coefficient: 0.02
        - var_name: under_3
          coefficient: 0.5
        - var_name: under_31
          coefficient: 0.02
        - var_name: under_33
          coefficient: 0.02
        - var_name: under_35
          coefficient: 0.02
        - var_name: under_37
          coefficient: 0.02
        - var_name: under_39
          coefficient: 4
        - var_name: under_41
          coefficient: 4
        - var_name: under_43
          coefficient: 4
        - var_name: under_45
          coefficient: 4
        - var_name: under_47
          coefficient: 4
        - var_name: under_5
          coefficient: 0.5
        - var_name: under_7
          coefficient: 0.5
        - var_name: under_9
          coefficient: 0.5
      sense: "="
      rhs: 0
  "_C4:":
    Standard:
      name: "_C4:"
      coefficients:
        - var_name: A_3
          coefficient: -1
        - var_name: A_7
          coefficient: 1
        - var_name: under_3
          coefficient: 1
      sense: ">="
      rhs: 0
  "_C5:":
    Standard:
      name: "_C5:"
      coefficients:
        - var_name: C_0
          coefficient: -1
        - var_name: C_7
          coefficient: 1
        - var_name: over_4
          coefficient: -1
      sense: "<="
      rhs: 0
  "_C6:":
    Standard:
      name: "_C6:"
      coefficients:
        - var_name: C_0
          coefficient: -1
        - var_name: C_7
          coefficient: 1
        - var_name: under_5
          coefficient: 1
      sense: ">="
      rhs: 0
  "_C7:":
    Standard:
      name: "_C7:"
      coefficients:
        - var_name: B_0
          coefficient: -1
        - var_name: B_3
          coefficient: 1
        - var_name: over_6
          coefficient: -1
      sense: "<="
      rhs: 0
  "_C8:":
    Standard:
      name: "_C8:"
      coefficients:
        - var_name: B_0
          coefficient: -1
        - var_name: B_3
          coefficient: 1
        - var_name: under_7
          coefficient: 1
      sense: ">="
      rhs: 0
  "_C9:":
    Standard:
      name: "_C9:"
      coefficients:
        - var_name: B_3
          coefficient: -1
        - var_name: B_7
          coefficient: 1
        - var_name: over_8
          coefficient: -1
      sense: "<="
      rhs: 0
